arrow/src/main/java/org/apache/iceberg/arrow/DictEncodedArrowConverter.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.iceberg.arrow.vectorized.ArrowVectorAccessor;
+import org.apache.iceberg.arrow.vectorized.VectorHolder;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/**
+ * This converts dictionary encoded arrow vectors to a correctly typed arrow vector.
+ */
+public class DictEncodedArrowConverter {
+
+  private DictEncodedArrowConverter() {
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static FieldVector toArrowVector(VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Preconditions.checkArgument(null != vectorHolder, "VectorHolder cannot be null");
+    Preconditions.checkArgument(null != accessor, "ArrowVectorAccessor cannot be null");
+    if (vectorHolder.isDictionaryEncoded()) {
+      if (Type.TypeID.DECIMAL.equals(vectorHolder.icebergType().typeId())) {
+        return toDecimalVector(vectorHolder, accessor);
+      } else if (Type.TypeID.TIMESTAMP.equals(vectorHolder.icebergType().typeId())) {
+        return toTimestampVector(vectorHolder, accessor);
+      } else if (Type.TypeID.LONG.equals(vectorHolder.icebergType().typeId())) {
+        return toBigIntVector(vectorHolder, accessor);
+      } else if (Type.TypeID.FLOAT.equals(vectorHolder.icebergType().typeId())) {
+        return toFloat4Vector(vectorHolder, accessor);
+      } else if (Type.TypeID.DOUBLE.equals(vectorHolder.icebergType().typeId())) {
+        return toFloat8Vector(vectorHolder, accessor);
+      } else if (Type.TypeID.STRING.equals(vectorHolder.icebergType().typeId())) {
+        return toVarCharVector(vectorHolder, accessor);
+      } else if (Type.TypeID.BINARY.equals(vectorHolder.icebergType().typeId())) {
+        return toVarBinaryVector(vectorHolder, accessor);
+      } else if (Type.TypeID.TIME.equals(vectorHolder.icebergType().typeId())) {
+        return toTimeMicroVector(vectorHolder, accessor);
+      }
+
+      throw new IllegalArgumentException(String.format("Cannot convert dict encoded field '%s' of type '%s' to Arrow " +
+              "vector as it is currently not supported",
+          vectorHolder.icebergField().name(), vectorHolder.icebergType().typeId()));
+    }
+
+    return vectorHolder.vector();
+  }
+
+  private static DecimalVector toDecimalVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    int precision = ((Types.DecimalType) vectorHolder.icebergType()).precision();
+    int scale = ((Types.DecimalType) vectorHolder.icebergType()).scale();
+
+    DecimalVector vector = new DecimalVector(
+        vectorHolder.vector().getName(),
+        ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+        vectorHolder.vector().getAllocator());
+
+    for (int i = 0; i < vectorHolder.vector().getValueCount(); i++) {
+      vector.setSafe(i, (BigDecimal) accessor.getDecimal(i, precision, scale));
```

Comment:
null value should be checked before setting it in the arrow vector. I think we need to add isNull(rowIndex) to the accessor and check vector.isNull(rowIndex).

Reply:
good point, do we need to do this for all the other vectors as well?

Reply:
currently I added it for those vectors that properly handle null when doing a `vector.getObject(i)`. All other access performs null checking when using `vector.get(i)` and requires `arrow.enable_null_check_for_get` to be set

Reply:
I think we should also have a follow-up PR where additional testing is done with null values, wdyt?

Reply:
Please do it for all vectors. 
getObject() is also expensive because it will create wrapper classes for primitive types. You can use vector.isSet(i) to check whether a value is set or not.
I'm fine with doing a follow-up PR to do additional testing.

Reply:
Nevermind my previous comment, the new change looks good to me.
For additional testing, a follow-up sounds good.

Reply:
> Please do it for all vectors.

sure thing, done.

Reply:
looks good